### PR TITLE
Signals now accept parameters

### PR DESCRIPTION
**Pull request by LeaoLuciano**
Project for testing: 
- [dodge-the-creeps-signals.zip](https://github.com/godot-rust/gdext/files/11507083/dodge-the-creeps-signals.zip)
- Added a signal named test in Player scene that is emitted when player dies
- This signal is connected in Main scene which calls test_function, printing player's death position

### CODE REVIEW

**Comment by Bromeon** [comment]
Thanks! 🙂 

Can you show some usage examples here (so people don't have to download the archive)?

**Comment by LeaoLuciano** [comment]
player.rs:
```rust
    #[signal]
    fn test(direction: Vector2);
``` 
main_scene.rs:
```rust
    #[func]
    pub fn new_game(&mut self) {
        ...
        let mut player = self.base.get_node_as::<player::Player>("Player");
        player.bind_mut().connect(
            "test".into(),
            Callable::from_object_method(self.base.get_node_as::<Self>("."), "test_function"),
            0,
        );
    }

    #[func]
    fn test_function(&self, direction: Vector2) {
        println!("Player died at: {}", direction);
    }

``` 

[Gravação de tela de 2023-05-18 08-31-44.webm](https://github.com/godot-rust/gdext/assets/56130740/e48c5344-b011-44d8-ace3-bb13c6498184)

**Comment by LeaoLuciano** [comment]
With 2 parameters:
player.rs:
```rust
    #[signal]
    fn test(direction: Vector2, distance: f32);


    #[func]
    fn on_player_body_entered(&mut self, _body: Gd<PhysicsBody2D>) {
        ...
        self.base.emit_signal("test".into(), &[pos.to_variant(), pos.length().to_variant()]);
    }
``` 
main_scene.rs:
```rust
    #[func]
    fn test_function(&self, direction: Vector2, distance: f32) {
        println!("Player died at: {} (distance {} from (0, 0))", direction, distance);
    }

``` 

[Gravação de tela de 2023-05-18 08-39-44.webm](https://github.com/godot-rust/gdext/assets/56130740/8e4605bf-b309-40e3-a2f4-2b6488da84fc)

**Comment by Bromeon** [comment]
bors try

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-279

File: itest/rust/src/object_test.rs [hunk 1]
```diff
@@ -856,3 +862,69 @@ fn double_use_reference() {
...
+        emitter.share().upcast::<Object>().connect(
+            signal_name.clone().into(),
+            receiver.callable(receiver_name),
+            0,
+        );
+
+        emitter
+            .share()
+            .upcast::<Object>()
+            .emit_signal(signal_name.into(), arg);
```
Comment (Bromeon) [inline itest/rust/src/object_test.rs:923]:
```suggestion
        emitter.base.emit_signal(signal_name.into(), arg);
```
It's in the same file and for a test, so encapsulation is not the highest priority.

File: itest/rust/src/object_test.rs [hunk 2]
```diff
@@ -856,3 +862,69 @@ fn double_use_reference() {
...
+    #[func]
+    fn receive_1_arg(&self, arg1: i64) {
+        self.used[1].set(true);
+        assert_eq!(arg1, 987);
+    }
+    #[func]
+    fn receive_2_arg(&self, arg1: Gd<Object>, arg2: GodotString) {
+        self.used[2].set(true);
+        assert_eq!(self.base.share(), arg1);
+        assert_eq!(SIGNAL_ARG_STRING, arg2.to_string());
```
Comment (Bromeon) [inline itest/rust/src/object_test.rs:889]: Would be nice to check the preconditions first and separate them with an empty line from the logic.
Same above.
```suggestion
        assert_eq!(self.base.share(), arg1);
        assert_eq!(SIGNAL_ARG_STRING, arg2.to_string());
        
        self.used[2].set(true);
```

File: itest/rust/src/object_test.rs [hunk 3]
```diff
@@ -856,3 +862,69 @@ fn double_use_reference() {
...
+    }
+}
+
+const SIGNAL_ARG_STRING: &str = "Signal string arg";
+
+#[itest]
+/// Test that godot can call a method that is connect with a signal
+fn signals() {
+    let emitter: Gd<SignalEmitter> = Gd::new_default();
+    let receiver: Gd<SignalReceiver> = Gd::new_default();
```
Comment (Bromeon) [inline itest/rust/src/object_test.rs:899]:
```suggestion
    let emitter = Gd::<SignalEmitter>::new_default();
    let receiver = Gd::<SignalReceiver>::new_default();
```
Turbofish! 🐟💨

File: itest/rust/src/object_test.rs [hunk 4]
```diff
@@ -856,3 +862,69 @@ fn double_use_reference() {
...
+        vec![Variant::from(987)],
+        vec![
+            Variant::from(receiver.share()),
+            Variant::from(SIGNAL_ARG_STRING),
+        ],
+    ];
+
+    for (i, arg) in args.iter().enumerate() {
+        let signal_name = format!("signal_{}_arg", i);
+        let receiver_name = format!("receive_{}_arg", i);
```
Comment (Bromeon) [inline itest/rust/src/object_test.rs:912]:
```suggestion
        let signal_name = format!("signal_{i}_arg");
        let receiver_name = format!("receive_{i}_arg");
```

File: itest/rust/src/object_test.rs [hunk 5]
```diff
@@ -856,3 +862,69 @@ fn double_use_reference() {
...
+
+        emitter.share().upcast::<Object>().connect(
```
Comment (Bromeon) [inline itest/rust/src/object_test.rs:914]:
```suggestion
        emitter.base.connect(
```

**Review by Bromeon — COMMENTED** [review]
Thanks a lot for the update! Added smaller comments. Since there is now quite a bit of signal functionality, you can also consider adding a new `signal_test.rs` file for these tests (don't forget license header). Or leave it as-is, if you prefer 🙂 

Could you integrate your changes into the 2nd commit?

**Comment by Bromeon** [comment]
bors try

**Review by Bromeon — APPROVED** [review]
Thanks a lot!

bors r+

**Pull request merged**